lara-typescript/src/interactive-api-client/index.ts

```diff
@@ -0,0 +1,344 @@
+              this.log("debug", "getInteractiveState listener called, client returned:", interactiveState);
+              this.setInteractiveState(interactiveState);
+            } else {
+              this.log("debug", "getInteractiveState listener called, no client callback found");
+            }
+          });
+
+          this.addListener("initInteractive",
+            // tslint:disable-next-line:max-line-length
+            (initMessage: LaraInteractiveApi.IInitInteractive<AuthoredState, InteractiveState, GlobalInteractiveState>) => {
+              if (this.options.onInitInteractive) {
+                this.log("debug", "initInteractive listener called, calling client callback");
+                this.options.onInitInteractive(initMessage);
+              } else {
+                this.log("debug", "initInteractive listener called, no client callback found");
+              }
+          });
+
+          this.addListener("loadInteractive", (interactiveState: InteractiveState | null) => {
+            if (this.options.onLoadInteractive) {
+              this.log("debug", "loadInteractive listener called, calling client callback");
+              this.options.onLoadInteractive(interactiveState);
+            } else {
+              this.log("debug", "loadInteractive listener called, no client callback found");
+            }
+          });
+
+          this.phone.initialize();
+        } else {
+          this.log("info", "#connect: this.phone already connected");
+        }
+        return true;
+      }
+      this.log("error", "#connect: not in iframe");
+      return false;
+    }
+
+    public disconnect() {
+      if (inIframe) {
+        if (this.phone) {
+          this.log("debug", "#disconnect: disconnecting");
+          this.phone.disconnect();
+          this.phone = undefined;
+        } else {
+          this.log("info", "#disconnect: this.phone already disconnected");
+        }
+        return true;
+      }
+      this.log("error", "#disconnect: not in iframe");
+      return false;
+    }
+
+    public get inIFrame() {
+      return inIframe;
+    }
+
+    public setLearnerUrl(url: string | null) {
+      return this.post("setLearnerUrl", url);
+    }
+
+    public setInteractiveState(interactiveState: InteractiveState | string | null) {
+      return this.post("interactiveState", interactiveState);
+    }
+
+    public setHeight(height: number | string) {
+      return this.post("height", height);
+    }
+
+    public setSupportedFeatures(supportedFeatures: ISupportedFeatures) {
+      return this.post("supportedFeatures", supportedFeatures);
+    }
+
+    public setNavigation(options: INavigationOptions) {
+      return this.post("navigation", options);
+    }
+
+    public setAuthoredState(authoredState: AuthoredState) {
+      return this.post("authoredState", authoredState);
+    }
+
+    public getAuthInfo(): Promise<IAuthInfo> {
+      return new Promise<IAuthInfo>((resolve, reject) => {
```

Comment:
This promise approach is great for these RPC type messages. However I think if we are going to formalize it like this. We need to add an id to `getAuthInfo` message which then is returned in the `authInfo` message. Otherwise it seems possible that this a client which asks for getAuthInfo twice could result in an unresolved promise.